src/shared/codelens/javaCodeLensProvider.ts

```diff
@@ -0,0 +1,206 @@
+ * @param symbol VS Code DocumentSymbol to analyze
+ */
+export function isValidMethodSignature(symbol: vscode.DocumentSymbol): boolean {
+    if (symbol.kind === vscode.SymbolKind.Method) {
+        // The `redhat.java` extension appears to strip a fair amount from this signature:
+        // from source function `public APIGatewayProxyResponseEvent handleRequest(final APIGatewayProxyRequestEvent input, final Context context)`
+        // redhat extension returns: symbol.name = `'handleRequest(APIGatewayProxyRequestEvent, Context)'`
+        const parametersArr = REGEXP_PARAMETERS.exec(symbol.name)
+        // reject if there are no parameters
+        if (!parametersArr) {
+            return false
+        }
+        // remove generics from parameter string so we can do a predictable split on comma
+        const strippedStr = stripGenericsFromParams(parametersArr[0])
+        const individualParams = strippedStr.split(',')
+        switch (individualParams.length) {
+            case 1:
+                return individualParams[0] === '()' ? false : true
+            case 2:
+                return lastParamIsContext(individualParams) || firstTwoParamsAreStreams(individualParams)
+            case 3:
+                return lastParamIsContext(individualParams) && firstTwoParamsAreStreams(individualParams)
+            default:
+                return false
+        }
+    }
+
+    return false
+}
+
+function lastParamIsContext(paramArr: string[]): boolean {
+    // TODO: handle different kinds of imported context objects in case user is importing a non-Lambda context
+    const lambdaContextType = /[\.\b]{0,1}Context\b/
+    return lambdaContextType.test(paramArr[paramArr.length - 1].valueOf().trimLeft())
+}
+
+function firstTwoParamsAreStreams(paramArr: string[]): boolean {
+    const inputStreamType = /[\.\b]{0,1}InputStream\b/
+    const outputStreamType = /[\.\b]{0,1}OutputStream\b/
+    return inputStreamType.test(paramArr[0].valueOf().trim()) && outputStreamType.test(paramArr[1].valueOf().trim())
+}
+
+/**
+ * Strips any generics from a string in order to ensure predictable commas for a string of parameters.
+ * e.g.: `'(Foo<Bar, Baz> x, Context y)' -> '(Foo x, Context y)'`
+ * Implements a fairly rough English-centric approximation of the Java identifier spec:
+ * * isJavaIdentifierStart(firstCharacter) is true: https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isJavaIdentifierStart(char)
+ * * all other characters are true for isJavaIdentifierPart: https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isJavaIdentifierPart(char)
+ *   * For now, cover this from an English-centric point of view (a-zA-Z) and add unicode ranges if necessary
+ *
+ * @param input String to remove generics from
+ */
+function stripGenericsFromParams(input: string): string {
+    const javaGenericIdentifierRegex = /(?:<{1}(?:\s*[a-zA-Z_$][a-zA-Z0-9_$]*[\s,]?)*>{1})/g
```

Comment:
lgtm analysis: "regular expression may cause exponential backtracking on strings containing many repetitions of '$'."

Is the `*` necessary? Maybe the warning can be avoided by limiting the matches instead of using `*`, e.g. (untested):

```suggestion
    const javaGenericIdentifierRegex = /(?:<{1}(?:\s*[a-zA-Z_$][a-zA-Z0-9_$]{,9}[\s,]?)*>{1})/g
```

Reply:
re: `{1}` I think that was for safety to make sure the `<` and `>` were included (this is a regex specifically to check English-language generics). I can try and retool if you'd like, this is mainly an adaptation of a similar one I did a long time ago for C# (there are slight differences in C# and Java object names)

Reply:
Resolved and checked with a combination of regexr.com and redos-checker.surge.sh .

Reply:
Ack, looks like there's an issue with eslint looking at this regex, specifically, it isn't understanding the possessive operator (`*+` at the end of the regex), which should resolve the LGTM issue. Will see if there's a way to satisfy both parties.